File: pkg/sharding/snowflake.go
package sharding

import (
	"sync"
	"time"
)

const epochMillis int64 = 1704067200000 // 2024-01-01T00:00:00Z

// 64-bit layout: 41 bits timestamp | 8 bits shard ID | 14 bits sequence.
// Modeled on Instagram's own published sharded-ID scheme, for the same
// reason they use it: the shard that minted an ID is recoverable forever
// with a bit-shift, no directory lookup, and it can never go stale. See
// doc/sharding.md.
const (
	sequenceBits   = 14
	shardIDBits    = 8
	shardIDShift   = sequenceBits
	timestampShift = sequenceBits + shardIDBits
	maxSequence    = (1 << sequenceBits) - 1

	// MaxShardID is the largest shard ID an 8-bit shard field can hold --
	// the hard ceiling on shard count this ID scheme supports without
	// changing the bit layout (see doc/sharding.md).
	MaxShardID = (1 << shardIDBits) - 1
)

// IDGenerator mints k-sortable, self-routing IDs for one specific shard.
// Each shard's writers run their own generator instance seeded with that
// shard's ID -- there is no coordination between shards because the
// shard ID is baked into every ID this generator produces.
type IDGenerator struct {
	mu            sync.Mutex
	shardID       int
	sequence      int64
	lastTimestamp int64
}

func NewIDGenerator(shardID int) *IDGenerator {
	if shardID < 0 || shardID > MaxShardID {
		panic("sharding: shardID out of range for an 8-bit ID field")
	}
	return &IDGenerator{shardID: shardID, lastTimestamp: -1}
}

func (g *IDGenerator) Next() int64 {
	g.mu.Lock()
	defer g.mu.Unlock()

	ts := time.Now().UnixMilli() - epochMillis
	if ts == g.lastTimestamp {
		g.sequence = (g.sequence + 1) & maxSequence
		if g.sequence == 0 {
			for ts <= g.lastTimestamp {
				ts = time.Now().UnixMilli() - epochMillis
			}
		}
	} else {
		g.sequence = 0
	}
	g.lastTimestamp = ts

	return (ts << timestampShift) | (int64(g.shardID) << shardIDShift) | g.sequence
}

// ExtractShardID recovers the shard that minted id -- a pure bit-shift,
// never a network call or a lookup table.
func ExtractShardID(id int64) int {
	return int(id>>shardIDShift) & MaxShardID
}

File: services/notification-service/src/storage/sharding/snowflake.ts
// Mirrors pkg/sharding's Go IDGenerator and com.feed.sharding's Java
// SnowflakeIdGenerator bit-for-bit: 41 bits timestamp | 8 bits shard ID |
// 14 bits sequence, same epoch. MUST use BigInt throughout -- these IDs
// exceed Number.MAX_SAFE_INTEGER (2^53-1), and silently doing this
// arithmetic in `number` would corrupt the low bits (the sequence and
// part of the shard ID) on every single ID. See doc/DESIGN.md.
const EPOCH_MILLIS = 1704067200000n; // 2024-01-01T00:00:00Z
const SEQUENCE_BITS = 14n;
const SHARD_ID_BITS = 8n;
const SHARD_ID_SHIFT = SEQUENCE_BITS;
const TIMESTAMP_SHIFT = SEQUENCE_BITS + SHARD_ID_BITS;
const MAX_SHARD_ID = (1n << SHARD_ID_BITS) - 1n;
const MAX_SEQUENCE = (1n << SEQUENCE_BITS) - 1n;

export class SnowflakeIdGenerator {
  private sequence = 0n;
  private lastTimestamp = -1n;

  constructor(private readonly shardId: number) {
    if (shardId < 0 || BigInt(shardId) > MAX_SHARD_ID) {
      throw new Error(`sharding: shardId out of range for an 8-bit ID field: ${shardId}`);
    }
  }

  next(): bigint {
    let ts = BigInt(Date.now()) - EPOCH_MILLIS;
    if (ts === this.lastTimestamp) {
      this.sequence = (this.sequence + 1n) & MAX_SEQUENCE;
      if (this.sequence === 0n) {
        while (ts <= this.lastTimestamp) {
          ts = BigInt(Date.now()) - EPOCH_MILLIS;
        }
      }
    } else {
      this.sequence = 0n;
    }
    this.lastTimestamp = ts;

    return (ts << TIMESTAMP_SHIFT) | (BigInt(this.shardId) << SHARD_ID_SHIFT) | this.sequence;
  }
}

/** Recovers the shard that minted id -- a bit-shift, never a lookup. */
export function extractShardId(id: bigint): number {
  return Number((id >> SHARD_ID_SHIFT) & MAX_SHARD_ID);
}
